### ISSUE TITLE
Store receipts in a table

### ISSUE THREAD

**Issue by kcole16** [near/near-explorer#218]
## Problem space

In Explorer, we don't have a way to know that some accounts were touched during cross-contract calls as we only track transactions, which has a pair of `signer_id` and `receiver_id`. However, contract execution may trigger cross-contract calls and touch "third-party" contracts.

E.g. #212, @evgenykuzyakov  thought that cross contract call was happening, no way to track. #337 was brought by @mattlockyer and points to the same issue of lack of receipts details to explore cross-contract calls.

## Solution

Track receipts (only exposed in TX detail as logs with status) in another table (second step)
  - Currently retrieving from RPC on TX detail page
  - Need to build receipt table somehow, associate with accounts

## Acceptance Criteria

* [x] A new `Receipt` model is defined to cover all the needs (indices and redundant columns to avoid expensive JOINs are in place)
* [x] Transaction status is saved into the transactions table resolving: https://github.com/nearprotocol/near-explorer/blob/6eae103df36df1b9918e7689aca2a3683e237eba/frontend/src/libraries/explorer-wamp/transactions.ts#L175-L179
* [x] "Ghost" accounts issue #223 is resolved
* [x] "Ghost" access keys should not be the case if we implement #162 first

**Comment by icerove** [comment]
@frol do you have any ideas what should we do for receipt table?

**Comment by frol** [comment]
Just for the context (it took me a few moments to recall what kind of "table" we are talking about here) it is a database table storing the receipts.

I have not dug into it before, so let me start with the facts about the current state.

There are receipt properties:
* `predecessor_id` (it could be either Transaction `signer_id` or intermediate contract's `account_id`)
* `receiver_id` (a receipt destination)
* `receipt_id:` (a unique id for the receipt; it is a hash)
* `receipt_type` one of:
  * `action`
    * `signer_id`
    * `signer_public_key`
    * `gas_price`
    * `output_data_receivers` is a list of:
       * `data_id` (a unique id for the data; it is a hash)
       * `receiver_id`
       * `input_data_ids` (a list of hashes)
       * `actions` a list of Actions (the same as in transactions)
  * `data`
    * `data_id` (a unique id for the data; it is a hash)
    * `data` (just a blob of data)

Currently, we can only query them with the `chunks` RPC API (block # 2871883 and shard # 7):

```
$ http post https://rpc.nearprotocol.com jsonrpc=2.0 id=dontcare method=chunk 'params:=[[2871883, 7]]'
```

The output:

```js
{
    "id": "dontcare",
    "jsonrpc": "2.0",
    "result": {
        "header": {
            "balance_burnt": "0",
            "chunk_hash": "Gx7CRg8goc6m3GBMyeufgSLaqkhH5HG2qEhxEHFYPW6J",
            "encoded_length": 191,
            "encoded_merkle_root": "ExpzxZNGvjMyRMphvDYY38XqD4rqJUiweDoXN6f8bQTY",
            "gas_limit": 1000000000000000,
            "gas_used": 937144500000,
            "height_created": 2871883,
            "height_included": 0,
            "outcome_root": "GYEAqSdzvjQZfd9KEF6LCiza5SjPAgsqYRKciGaMUd4s",
            "outgoing_receipts_root": "5w2Cx8paWdCV4TCCwoMcc3NTpgAeBvNoGjCkGkW1pHqX",
            "prev_block_hash": "gSchiKdxvH7MtG9zqZsXLpTUpVRBRUUosqz59n1T6PB",
            "prev_state_root": "GqRtmo1upVD94xt3od9ak9goULxwAQbDrTjCou7VXo3j",
            "rent_paid": "1178834",
            "shard_id": 7,
            "signature": "ed25519:4fqeabHFfgGdV6m3ja5z42maWZ6dqnpqXeNekz2fEcCyUMGj4HvB1AA5YzbwR8AhyWeQherdLrcram78D5monp11",
            "tx_root": "11111111111111111111111111111111",
            "validator_proposals": [],
            "validator_reward": "4685722500000000"
        },
        "receipts": [
            {
                "predecessor_id": "test",
                "receipt": {
                    "Action": {
                        "actions": [
                            "CreateAccount",
                            {
                                "Transfer": {
                                    "deposit": "10000001000000000000000000"
                                }
                            },
                            {
                                "AddKey": {
                                    "access_key": {
                                        "nonce": 0,
                                        "permission": "FullAccess"
                                    },
                                    "public_key": "ed25519:36Gmm6qJ4a78qTknvqLUPXoGp5KLeAvDDTM194ngkFua"
                                }
                            }
                        ],
                        "gas_price": "5000",
                        "input_data_ids": [],
                        "output_data_receivers": [],
                        "signer_id": "test",
                        "signer_public_key": "ed25519:oNCFEmRotRHTySKqmAwifNZ8VRpYS973p8cL61y7eiE"
                    }
                },
                "receipt_id": "AhbJN8J4Nk5aFxUptkkrJ5EJWsTiEQtjBTHvGt9sPFzW",
                "receiver_id": "victorvw"
            }
        ],
        "transactions": []
    }
}
```

Also, there is a receipt outcome (same as we have for transactions):
* `block_hash` (the block where this receipt was included)
* `gas_burnt`
* `logs` (a list of strings)
* `receipt_ids` (a list of follow-up receipt ids)
* `status` - it is an enum, but it is serialized as an object with the following possible keys:
  * `Failure`
  * `SuccessValue` (the final action succeeded and returned some value or an empty string; the value is the value encoded as base64)
  * `SuccessReceiptId` (the final action of the receipt returned a promise; the value is the `receipt_id` of the generated receipt)
* `proof`, example:
  
  ```js
  "proof": [
                {
                    "direction": "Right",
                    "hash": "45LBFtJSxFcVSWiiamY11GZ9nq1Cmcfc5xueD49GZDZz"
                }
            ]
  ```

We get the receipt outcomes with the `tx` RPC API:

```
$ http post https://rpc.nearprotocol.com jsonrpc=2.0 id=dontcare method=tx 'params:=["EzvvJEqdxKA62oAjG32y5herDtRGYEdmZXHDDyxorJ48", "victorvw"]'
```

The output:

```js
{
    "id": "dontcare",
    "jsonrpc": "2.0",
    "result": {
        "receipts_outcome": [
            {
                "block_hash": "9YGAYkA4mhafu1RWFKpQvVJMWdFeaCM6fW6Jqair2h52",
                "id": "AhbJN8J4Nk5aFxUptkkrJ5EJWsTiEQtjBTHvGt9sPFzW",
                "outcome": {
                    "gas_burnt": 937144500000,
                    "logs": [],
                    "receipt_ids": [],
                    "status": {
                        "SuccessValue": ""
                    }
                },
                "proof": []
            }
        ],
        "status": {
            "SuccessValue": ""
        },
        "transaction": {
            "actions": [
                "CreateAccount",
                {
                    "Transfer": {
                        "deposit": "10000001000000000000000000"
                    }
                },
                {
                    "AddKey": {
                        "access_key": {
                            "nonce": 0,
                            "permission": "FullAccess"
                        },
                        "public_key": "ed25519:36Gmm6qJ4a78qTknvqLUPXoGp5KLeAvDDTM194ngkFua"
                    }
                }
            ],
            "hash": "EzvvJEqdxKA62oAjG32y5herDtRGYEdmZXHDDyxorJ48",
            "nonce": 10270,
            "public_key": "ed25519:oNCFEmRotRHTySKqmAwifNZ8VRpYS973p8cL61y7eiE",
            "receiver_id": "victorvw",
            "signature": "ed25519:46WsywQYe31isWcHASrnR2pYCja4Mtjzs4n87isrTHWyDdX5uwrUaUn4SbhY9BMwWKvWwdcuyCjMsXdkXdPJv8ko",
            "signer_id": "test"
        },
        "transaction_outcome": {
            "block_hash": "gSchiKdxvH7MtG9zqZsXLpTUpVRBRUUosqz59n1T6PB",
            "id": "EzvvJEqdxKA62oAjG32y5herDtRGYEdmZXHDDyxorJ48",
            "outcome": {
                "gas_burnt": 937144500000,
                "logs": [],
                "receipt_ids": [
                    "AhbJN8J4Nk5aFxUptkkrJ5EJWsTiEQtjBTHvGt9sPFzW"
                ],
                "status": {
                    "SuccessReceiptId": "AhbJN8J4Nk5aFxUptkkrJ5EJWsTiEQtjBTHvGt9sPFzW"
                }
            },
            "proof": []
        }
    }
}
```

To be continued...

**Comment by frol** [comment]
I would start with what users may want to know:

**Q: Which transactions (including cross-contract calls) touched the given account?**

A: To know that, we need to store `predecessor_id` and `receiver_id`, and a reference to the transaction (`transaction_hash`).

**Q: What is the execution "graph" of the receipts?**

A: Given that a single receipt can produce many new receipts and then gather them together, we need some way to know the parent/child relationship, and instead of having a list of child receipts, we can have a `parent_receipt_id` (NULL for the first receipt), as it just how we normalize the data.

**Q: What are the input and the result of the receipt?**

A: Input is an action (the same type as we have for transactions), and the result is the outcome status. We should take some time thinking about these receipt actions: can/should we reuse the transaction actions here?

**Q: What went wrong with the execution?**

A: In addition to the input and the results, we also need `logs` and intermediate values (arguments and return values). I would prefer not to store those in Explorer DB, so we just re-query that on the Transaction Details page on demand (e.g. when a user digs into the receipt details [NOTE: we need RPC API for that]).

*It needs some better thought, but it is a starting point. I would like to invite more people to express their opinions on what they want to see on Explorer in regards to the receipts /cc @fckt @k06a @willemneal @AnaisUrlichs @ilblackdragon @bowenwang1996 @Kouprin @janedegtiareva @vgrichina*

**Comment by frol** [comment]
The core of this issue is resolved with #236.

Transactions that are touching a given account are now properly displayed on the Account details page as of #506.

The UI for Receipts on Transaction details page is tracked under #512.
